graql/reasoner/query/DisjunctiveQuery.java

```diff
@@ -0,0 +1,304 @@
+                    }));
+            LazyMergingStream<ConceptMap> mergedStreams = new LazyMergingStream<>(answerStreams);
+            return mergedStreams.flatStream();
+        } else {
+            return new ResolutionIterator(this, subGoals, context().queryCache()).hasStream();
+        }
+    }
+
+    @Override
+    ResolvableQuery inferTypes() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(CompositeQuery::inferTypes).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery constantValuePredicateQuery() {
+        return new DisjunctiveQuery(
+                getClauses().stream().map(CompositeQuery::constantValuePredicateQuery).collect(Collectors.toSet()),
+                getBindingVars(),
+                traversalExecutor,
+                context()
+        );
+    }
+
+    @Override
+    ResolvableQuery copy() {
+        return new DisjunctiveQuery(
+                new HashSet<>(getClauses()),
+                getBindingVars(),
+                traversalExecutor,
+                context());
+    }
+
+    @Override
+    public boolean isAtomic() {
+        // TODO unclear of the meaning of atomicity in this case
+        return getClauses().stream().allMatch(CompositeQuery::isAtomic);
+    }
+
+    @Override
+    boolean isEquivalent(ResolvableQuery q) {
+        DisjunctiveQuery that = q.asDisjunctive();
+        return that.getClauses().size() == getClauses().size()
+                && getClauses().stream().allMatch(clause -> that.getClauses().stream().anyMatch(clause::isEquivalent));
+    }
+
+    @Override
+    public String toString(){
+        return getPattern().toString();
+    }
+
+    @Override
+    public ReasonerQuery conjunction(ReasonerQuery q) {
+        throw new UnsupportedOperationException();
+//        return new DisjunctiveQuery(
+//                getClauses().stream()
+//                        .map(CompositeQuery::getPattern)
+//                        .map(p -> Graql.and(
+//                                Sets.union(
+//                                        p.getPatterns(), getPattern().getPatterns()
+//                                )
+//                        ))
+//                        .map(p -> new CompositeQuery(
+//                                p, traversalExecutor, context()
+//                        ))
+//                        .collect(Collectors.toSet()),
+//                traversalExecutor,
+//                context()
+//        );
+    }
+
+    @Override
+    public void checkValid() {
+        clauses.forEach(CompositeQuery::checkValid);
+    }
+
+    @Override
+    public Set<Variable> getVarNames() {
+        return bindingVars;
+    }
+
+    @Override
+    public Set<Atomic> getAtoms() {
+        return getClauses().stream().flatMap(c -> c.getAtoms().stream()).collect(Collectors.toSet());
```

Comment:
@kasper-piskorski also here

Reply:
You could either return all atoms or throw. Throwing might help highlighting unexpected execution paths. If you want to actually use it then return all atoms.